feat(LabelGroup): add editable label examples

kmcfaul opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Closes #7351

@mcarrano LMK if there are any adjustments I should make to the new examples and whether these should live in a demo page or remain on the examples page. I followed the scenarios pretty closely but didn't cover every shown variation (re: label colors, within groups or out, etc).

Added a new property `addLabel` to `LabelGroup` which puts the label within the label group but not included in the overflow list.

patternfly-build commented:
Preview: https://patternfly-react-pr-7515.surge.sh

mcarrano commented:
This is looking good @kmcfaul . Just a couple of comments...

* When using the dropdown to pick from a list of labels, I would not expect the resulting new label to be editable. The "pick" use case is for situations where the label name is constrained to some pre-existing list. 
* For the modal example, seems like the resulting label is editable even if I set Editable=No.

I think that you raise an interesting question about whether any of these examples are really demos. The Modal example definitively feels like it should be a demo since the content of the modal could be anything. I'd like to not imply that we must expose all of the options. Maybe the other two are OK as examples? Somehow we should make it clear that you can customize the Add function to be what you need it to be.

@mmenestr perhaps also take a look at this.

mmenestr commented:
This is looking good!
Just a few comments:
• For editable labels with add button, I would expect that when I click into the edit the label, the text doesn't change automatically to something else - I think it should stay saying "New label" until I manually type something else in? I also think I would expect it to remain editable after I've renamed it something other than "New label". @mcarrano please tell me if that's not expected...
• Agreed with Matt that for the dropdown one it probably shouldn't be editable post dropdown selection!
• For the modal one, when I select everything and submit, everything looks great and works as expected, but once I go in and try to edit the label that was created, it reverts back to what seems to be the default label? (Ie: A blue filled label with the words "Label 1"). I would expect it to retain the characteristics I had selected in the modal!

mcarrano commented:
> I would expect it to remain editable after I've renamed it something other than "New label". @mcarrano please tell me if that's not expected...

Yes, I agree with that.

tlabaj commented:
Other than the example where the control is alabel

> This is looking good @kmcfaul . Just a couple of comments...
> 
> * When using the dropdown to pick from a list of labels, I would not expect the resulting new label to be editable. The "pick" use case is for situations where the label name is constrained to some pre-existing list.
> * For the modal example, seems like the resulting label is editable even if I set Editable=No.
> 
> I think that you raise an interesting question about whether any of these examples are really demos. The Modal example definitively feels like it should be a demo since the content of the modal could be anything. I'd like to not imply that we must expose all of the options. Maybe the other two are OK as examples? Somehow we should make it clear that you can customize the Add function to be what you need it to be.
> 
> @mmenestr perhaps also take a look at this.

Other than the example that uses a label as a control, the other two feel like demos.  I wonder if we should move the dropdown and modal examples to demos.  We could add some text to the label example saying see demos for more advance/other ways to add a label?

tlabaj commented on packages/react-core/src/components/LabelGroup/LabelGroup.tsx:
```diff
@@ -58,6 +58,8 @@ export interface LabelGroupProps extends React.HTMLProps<HTMLUListElement> {
   /** @beta Additional props passed to the editable textarea. */
   editableTextAreaProps?: any;
+  /** @beta Control for adding new labels */
+  addLabel?: React.ReactNode;
```
Maybe `addLabelControl` or `addLabelAction` for a name would be more descriptive of prop intent?

srambach commented:
If I delete all the labels in the group, the "add label" and the entire label group also disappears. Should it stick around?

srambach commented:
Also, two other things I notice.
- If you created an editable label, then focus on it with mouse or keyboard to edit it, it always changes to be "Label 1". I'd expect it to retain its current value.
- In the modal, if you use the keyboard to navigate, selecting something from a dropdown throws focus to the close button. I'd expect it to go back to the dropdown.

mcarrano commented:
> If I delete all the labels in the group, the "add label" and the entire label group also disappears. Should it stick around?

That's a great catch @srambach . This was not covered in the design spec, but the way I'd propose this to work is that when the last label is removed, some placeholder text is added to the group that defaults to something like, "Add a new label," and the Add label button remains active so the user can add more labels to the group. What do you think @kmcfaul @mmenestr ?

Also, in reviewing this again, I noticed a small core issue that slipped through the cracks. Per our content guidelines, the button label should be "Add label" and not "Add Label."

mmenestr commented:
Idk that having an "add a new label" text would even be needed ? I think just making sure the "Add label" button never disappears is enough!

kmcfaul commented:
Updates: [@mcarrano @mmenestr]
- Removed editable flag for new options added by dropdown
- Fixed add via modal not respecting editable option
- Fixed label text changing for newly added labels
- Fixed label not being editable after editing
- Fixed label changing props after editing
- Renamed `addLabel` to `addLabelControl`
- Moved dropdown + modal add to demos; added description to point to demos
- Prevented label group disappearing when number of labels is 0 & `addLabelControl` is defined
- Updated add label button content to "Add label"

@evwilkin Is it possible to put demo `md` files inside the component folder? I was only able to get the docs build to pick up the new file when it is located in the demo folder (for now I have the `md` there and the code files are still in the component folder).

@srambach I'm looking into the focus issue, I believe it has to do with the fact the toggle is a `div` wrapping the `Label` rather than a typical `MenuToggle`-like component it was built to expect.

mcarrano reviewed (changes requested):
@kmcfaul This looks great now. I just stumbled over one odd bug. If I add multiple labels with the same text (i.e. "Label text 1), and then I remove one of them, it removes all the labels with the same text.

mmenestr commented:
Looks really good!! Just have a few more random things:

1. One random comment - and i'm not sure what other people would think, but: If I go into edit a label, and then open the dropdown before finishing my edit, the label I was editing goes into a weird in between state...

https://user-images.githubusercontent.com/51165119/172721100-e62ddf33-967b-4f58-a855-46533b4350a0.mov

But what I would expect to have happen is that if the label is mid edit when I click on the dropdown, it would revert back to its default state as soon as I click the dropdown. 

2. In the modal example, I'm wondering if the fields should be required or not? If I don't select anything, I notice that it just defaults to "No" on everything.. I don't think that's wrong per say, but seeing as the form asks the user to select a yes or no, It feels weird not to force them to select something? @mcarrano what are your thoughts on that?

Also as a side note (and might be something that needs to be fixed outside of this issue) but I noticed that when I try to edit a label by highlighting all the text (to eventually delete it), if I move my cursor past the label (like left of the label while I highlight) it comes out of the edit mode when I release my click - which is an annoying interaction. I feel like it should stay in the edit mode until I click away from it, not when I release a click on highlight. Let me know if that doesn't make sense 

https://user-images.githubusercontent.com/51165119/172721513-299c03a1-26fb-4a72-b9fe-0470ef251f6d.mov

mcarrano commented:
> In the modal example, I'm wondering if the fields should be required or not? If I don't select anything, I notice that it just defaults to "No" on everything.. I don't think that's wrong per say, but seeing as the form asks the user to select a yes or no, It feels weird not to force them to select something? @mcarrano what are your thoughts on that?

Since this is only a demo, I'm not too concerned with whether the fields are indicated as required or not. In a real use case I would expect a dialog that presents more application specific options. As for the yes/no options, it's generally good practice to have a default selection when using radio buttons so if the default is "no" then the radio buttons should be set to indicate that.

kmcfaul commented:
- Added default values for yes/no modal options and filled/outline modal option
- Adjusted dropdown example to not prevent event propagation on toggle click (which prevented editable label handler). This now has race condition between the label turning off the editable state and dropdown opening, where the menu opens off-center. I'm not sure how best to address this.
- Adjusted examples to close based on an ID to prevent matching labels being removed all at once.

tlabaj commented:
> @kmcfaul these two lines in the `source` file in react-docs define the file structure where the PF theme is looking for .md files and which tabs it's mapping them to:
> 
> * https://github.com/patternfly/patternfly-react/blob/main/packages/react-docs/patternfly-docs.source.js#L39-L40
> 
> Based on this you could move the demo code into a new `demos` directory like one of these options:
> 
> * `LabelGroup/examples/demos/[file-here].md`
> * `LabelGroup/demos/[file-here].md`
> 
> For consistency I'd check with @tlabaj if you want to do this as I don't believe we organize any other existing demos this way currently.

@kmcfaul currently all our demos live in `patternfly-react/packages/react-core/src/demos`.  Probably left over from when the demos where in a separate section on the website.  We can discuss whether it makes more sense to restructure where we put the demos in the code.  I agree that it may make more sense for them to live with the component.

mcarrano reviewed (approved):
Looks good to me. Thanks @kmcfaul !

mmenestr reviewed (approved):
Looks great to me too !! Thanks @kmcfaul

srambach commented:
A couple of keyboard things -
- When I edit a label and hit return to persist, the focus is not shown. Tab puts it back onto the control but if it could be there immediately that would be better.
- If I add some labels, then add but just press enter on the dialog, all the ones I added disappear
- I'm having trouble consistently reproducing it, but a couple of times, using just keyboard, I've added some editable labels, edited one or two, and hit one of the delete buttons and all the ones I created disappeared.

wise-king-sullyman reviewed (approved):
Looks great!

thatblindgeye commented on packages/react-core/src/components/LabelGroup/examples/LabelGroupEditableAdd.tsx:
```diff
@@ -0,0 +1,74 @@
+    setLabels(labels.filter((l: any) => l.id !== labelId));
+  };
+
+  const onEdit = (nextText: string, index: number) => {
+    const copy = [...labels];
+    copy[index] = { name: nextText, props: labels[index].props };
+    setLabels(copy);
+  };
```
For the bug shown in the video above, it looks like the issue occurs here; the `label.id` ends up getting removed after editing.

kmcfaul replied:
Ah good catch, I missed preserving that when I updated everything to use an id instead of name.

thatblindgeye reviewed (commented):
This looks great! I believe I'm also seeing the same behavior that @srambach is seeing, albeit this video is using mouse (I was able to recreate the behavior with keyboard only, though):

https://user-images.githubusercontent.com/70952936/173415160-4e2db066-ffa5-44e1-9aa3-1ea87cb72be0.mov

I would also agree with focus being placed on the editable label after editing has concluded with the Enter key, though since this component is still beta I wonder if addressing a11y concerns in a separate issue would be better.

kmcfaul commented:
I will open a follow up to address keyboard navigation for editable labels.

thatblindgeye reviewed (approved):
Looks good! 🎉

tlabaj commented:
> @srambach The `onClick` event uses the position of the cursor on mouse-up so if the cursor is outside when released it will be detected as an outside click (cancelling the edit). You can properly select the text as long as the cursor is still within the input, or by clicking several times to select all.
> 
> Would we rather have editing be active using `onMouseDown`?

cc @mmenestr

mmenestr commented:
If i'm understanding this correctly, I would vote MouseUp for editing being active!

kmcfaul commented:
> If i'm understanding this correctly, I would vote MouseUp for editing being active!

That's currently what is happening (because the click event position is the MouseUp position). The caveat is that when you click and drag, to highlight text, you potentially will trigger a cancelled edit if you release outside the input.

thatblindgeye commented:
@kmcfaul @mmenestr If using MouseDown would resolve that issue without causing any other major issues, I think it would be more beneficial to use that instead. One benefit could be for users who may not have the accuracy to keep the mouse within the input the entire time when selecting that text.

srambach commented:
> One benefit could be for users who may not have the accuracy to keep the mouse within the input the entire time when selecting that text.

+1

kmcfaul commented:
@srambach @thatblindgeye @tlabaj @mmenestr 

Editing will now activate `onMouseDown` for mouse interaction, keyboard interaction will be the same. There shouldn't be a major difference in feel for the mouse interaction, just the editable field activating a fraction earlier.

srambach reviewed (approved):
Yay! Much better feel now when selecting text to edit. Once the focus issues are addressed later, this will be really slick. Thanks for addressing the comments. :)
